notebooks/Learning_Rate_Scheduler.ipynb

```diff
@@ -0,0 +1,3350 @@
+   "execution_count": 11,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def prepare_data():\n",
+    "    # Load the dataset\n",
+    "    data = load_breast_cancer()\n",
+    "    X, y = data.data, data.target\n",
+    "    \n",
+    "    # Split into train and test sets\n",
+    "    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)\n",
+    "    \n",
+    "    # Standardize the features\n",
+    "    scaler = StandardScaler()\n",
+    "    X_train_scaled = scaler.fit_transform(X_train).astype(np.float32)\n",
+    "    X_test_scaled = scaler.transform(X_test).astype(np.float32)\n",
+    "    \n",
+    "    # Reshape the labels for compatibility\n",
+    "    y_train = y_train.astype(np.float32).reshape(-1, 1)\n",
+    "    y_test = y_test.astype(np.float32).reshape(-1, 1)\n",
+    "    \n",
+    "    return X_train_scaled, X_test_scaled, y_train, y_test\n",
+    " "
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Neural Net Parameters\n",
+    "\n",
+    "The BreastCancerNet is a neural network designed for binary classification tasks. It consists of an input layer, two hidden layers with ReLU activation functions, and a single output layer. The architecture is parameterized to allow flexibility in adjusting the input and hidden layer dimensions."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 12,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "class BreastCancerNet(nn.Module):\n",
+    "    def __init__(self, input_dim=30, hidden_dim=64):\n",
+    "        super(BreastCancerNet, self).__init__()\n",
+    "        self.fc1 = nn.Linear(input_dim, hidden_dim)\n",
+    "        self.relu = nn.ReLU()\n",
+    "        self.fc2 = nn.Linear(hidden_dim, 1)\n",
+    "\n",
+    "    def forward(self, x):\n",
+    "        x = self.fc1(x)\n",
+    "        x = self.relu(x)\n",
+    "        x = self.fc2(x)\n",
+    "        return x\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Learning Rate Scheduler Parameters \n",
+    "\n",
+    "1) StepLR:\n",
+    "\n",
+    "- Reduces the learning rate by a factor (gamma=0.3) every 100 epochs (step_size=10).\n",
+    "- Useful for steady, predictable learning rate decay.\n",
+    "\n",
+    "2) ReduceLROnPlateau:\n",
+    "\n",
+    "- Reduces the learning rate dynamically when the model's performance (e.g., loss) plateaus.\n",
+    "- Adjusts by a factor (factor=0.7) after 5 epochs of no improvement (patience=5).\n",
+    "- Ideal for tasks where loss stagnation indicates the need for smaller learning rates.\n",
+    "\n",
+    "3) CosineAnnealing:\n",
+    "\n",
+    "- Reduces the learning rate in a cosine curve over 10 epochs (T_max=10).\n",
+    "- Periodically resets the learning rate, promoting exploration of the loss landscape.\n",
+    "\n",
+    "We will now simulate training the neural network with different LR Schedulers."
```

Comment:
This is not a "simulation", right? We really do the training. Let's rephrase this.

Reply:
Updated as discussed.